### PR TITLE
PR for #18

### PR DESCRIPTION
**Pull request by MaSven**
This is the PR to track progress of Issue #18

### CODE REVIEW

**Comment by marko-radosavljevic** [comment]
I like what I see so far! 
Handling is even rougher than I expected, and you are doing even more than I asked. Great work. 

I should update the issue to include:
0. Avoid empty catch blocks at all costs, if an empty catch block is intentional, make it clear with variable name and comment.
1. Close every resource with _try-with-resources_.
2. Avoiding `null` checks that will hide bugs. Fail fast. `NullPointerException` is much more preferred to method failing to execute silently.
3. When there is new exception being thrown inside the catch block I prefer sending the original exception with the new one. More precisely, constructor `Exception(String message, Throwable cause)`.
4. Every exception should have message describing what caused it.
5. Fail even faster.

**Comment by MaSven** [comment]
What I also encountered, is that a lot of files are formatted not correctly. Maybe you should also open a new Issue to suggest using spotless or something similar so everyone does have the same formatting. Could also be a good first issue.

**Comment by marko-radosavljevic** [comment]
Yup we will fix that in separate issue. It's kinda funny, read end of the readme.md.
I will create issues like that when I write contributing.md with instructions on styles and stuff like that. For now don't auto-format.

**Comment by anirbandey303** [comment]
Is this issue resolved? I'll work on a different one then.

**Comment by marko-radosavljevic** [comment]
@anirbandey303 this is PR, but yeah, this issue is assigned to @MaSven 
We will have many new issues soon, so stay tuned! ^^

File: src/main/java/de/myopenlab/update/DeletePackages.java [hunk 1]
```diff
@@ -1,44 +1,54 @@
...
+            deleteEveryFolderFromList();
 
             Thread.sleep(1000);
             owner.log("finished");
             //owner.initTable1();
             owner.initTable2();
             owner.owner.reinitPackage();
         } catch (InterruptedException ex) {
-            Logger.getLogger(DeletePackages.class.getName()).log(Level.SEVERE, null, ex);
+            Logger.error(ex,"Exception while trying to sleep in thread");
```
Comment (marko-radosavljevic) [inline src/main/java/de/myopenlab/update/DeletePackages.java:29]: Maybe add `Thread.currentThread().interrupt();` here?

File: src/main/java/de/myopenlab/update/Tools2.java [hunk 2]
```diff
@@ -44,23 +43,45 @@ public static void copy(InputStream input, OutputStream output, int bufferSize)
...
+     * @param connection   The connection from where we want to download the package
+     * @throws PackageTransportationException If the copy failed
+     */
+    private static void downloadPackage(String destfilename, HttpURLConnection connection) throws PackageTransportationException {
+        try (InputStream in = connection.getInputStream();
+             FileOutputStream out = new FileOutputStream(destfilename)) {
+            copy(in, out, 1024);
+
+        } catch (IOException copyException) {
+            throw new PackageTransportationException("Error. Trying to copy package failed", copyException);
```
Comment (marko-radosavljevic) [inline src/main/java/de/myopenlab/update/Tools2.java:84]: :+1:

File: src/main/java/de/myopenlab/update/Tools2.java [hunk 3]
```diff
@@ -44,23 +43,45 @@ public static void copy(InputStream input, OutputStream output, int bufferSize)
...
+    }
+
+    /**
+     * Download the package and copy its content to desfilename
+     *
+     * @param destfilename Where should the package get downlaoded
+     * @param connection   The connection from where we want to download the package
+     * @throws PackageTransportationException If the copy failed
+     */
+    private static void downloadPackage(String destfilename, HttpURLConnection connection) throws PackageTransportationException {
```
Comment (marko-radosavljevic) [inline src/main/java/de/myopenlab/update/Tools2.java:78]: :heart:

File: src/main/java/de/myopenlab/update/frmUpdate.java [hunk 4]
```diff
@@ -77,59 +60,71 @@
...
 
-        //con.setRequestProperty("User-Agent", USER_AGENT);
-        int responseCode = con.getResponseCode();
-        System.out.println("\nSending 'GET' request to URL : " + url);
-        System.out.println("Response Code : " + responseCode);
+            StringBuilder response = readBufferToString(con);
 
+            return response.toString();
+        } catch (IOException urlException) {
+            throw new PackageTransportationException(urlException);
```
Comment (marko-radosavljevic) [inline src/main/java/de/myopenlab/update/frmUpdate.java:89]: Maybe add causal exception in constructor?

File: src/main/java/de/myopenlab/update/frmUpdate.java [hunk 5]
```diff
@@ -77,59 +60,71 @@
...
         StringBuilder response;
         try (BufferedReader in = new BufferedReader(new InputStreamReader(con.getInputStream()))) {
             String inputLine;
             response = new StringBuilder();
             while ((inputLine = in.readLine()) != null) {
-                response.append(inputLine + "\n");
+                response.append(inputLine).append("\n");
             }
+        } catch (IOException ioException) {
+            throw new PackageTransportationException(ioException);
```
Comment (marko-radosavljevic) [inline src/main/java/de/myopenlab/update/frmUpdate.java:103]: Maybe add causal exception in constructor?

File: src/main/java/de/myopenlab/update/frmUpdate.java [hunk 6]
```diff
@@ -1322,17 +1275,50 @@ private void jMenuItemUploadPackageActionPerformed(java.awt.event.ActionEvent ev
...
+                Logger.debug(str);
                 Tools.showMessage(this, "" + str);
             }
         } catch (IOException ex) {
-            Logger.getLogger(frmUpdate.class.getName()).log(Level.SEVERE, null, ex);
+            Logger.error(ex,"Error. Tried to read or write");
         }
     }//GEN-LAST:event_jMenuItemUploadPackageActionPerformed
 
+    private void writeFile(String entry_name, String type, String charset, File binaryFile, String boundary, String CRLF, URLConnection connection) throws IOException {
```
Comment (marko-radosavljevic) [inline src/main/java/de/myopenlab/update/frmUpdate.java:1289]: Love the extractions, you refactored almost everything  ^^ :heart:

**Review by marko-radosavljevic — COMMENTED** [review]
This took longer than I expected.. Great work! ^^
I want this merged asap, all other work is depending on this. We will finish 2nd part later.

Few things I noted after this:

1.  We should have much smaller and more contained commits, so it doesn't break workflow for other devs, and it's easier to thoroughly review, and also easier to grep the exact change you want from git logs. (we underestimated the magnitude of exception handling issues).
2. We should have a class with an error message as a field, so we can have consistent error messages across the codebase, also an object we can poke. All error messages should be enumerated.

**Comment by MaSven** [comment]
Iam on vacation till Sunday evening. I pushed everything I still had left. So this PR will stale till Sunday

**Comment by marko-radosavljevic** [comment]
We can't wait that long, this issue is stopping everything else. But thanks for notifying me! ^^

I will fix these minor things and push myself. Thanks

**Comment by MaSven** [comment]
OK I will change the PR from draft to ready so you can merge it later. On Sunday I can work again at this or another issue.

**Pull request closed without merging**